File: ts-server/src/tools/list-todos.ts
import { todos } from "../types.js";

export async function listTodos(completed?: boolean) {
  let filteredTodos = Object.values(todos);

  if (completed !== undefined) {
    filteredTodos = filteredTodos.filter((t) => t.completed === completed);
  }

  if (filteredTodos.length === 0) {
    return {
      content: [
        {
          type: "text" as const,
          text: "No todos found.",
        },
      ],
    };
  }

  const todoList = filteredTodos
    .map((t) => {
      const status = t.completed ? "✓" : " ";
      const desc = t.description ? ` - ${t.description}` : "";
      return `[${status}] ${t.title}${desc} (ID: ${t.id})`;
    })
    .join("\n");

  return {
    content: [
      {
        type: "text" as const,
        text: `Todos:\n${todoList}`,
      },
    ],
  };
}

File: python-server/src/tools/list_todos.py
from ..types import todos


async def list_todos(completed: bool | None = None) -> dict:
    filtered_todos = list(todos.values())

    if completed is not None:
        filtered_todos = [t for t in filtered_todos if t.completed == completed]

    if len(filtered_todos) == 0:
        return {
            "content": [
                {
                    "type": "text",
                    "text": "No todos found.",
                },
            ],
        }

    todo_list = "\n".join(
        f"[{'✓' if t.completed else ' '}] {t.title}{f' - {t.description}' if t.description else ''} (ID: {t.id})"
        for t in filtered_todos
    )

    return {
        "content": [
            {
                "type": "text",
                "text": f"Todos:\n{todo_list}",
            },
        ],
    }
